numpy/random/src/philox/philox.h

```diff
@@ -25,10 +25,17 @@ _philox4x64bumpkey(struct r123array2x64 key) {
   return key;
 }
 
+/* Prefer uint128 if available: GCC, clang, ICC */
+#ifdef __SIZEOF_INT128__
+static NPY_INLINE uint64_t mulhilo64(uint64_t a, uint64_t b, uint64_t *hip) {
+  __uint128_t product = ((__uint128_t)a) * ((__uint128_t)b);
+  *hip = product >> 64;
+  return (uint64_t)product;
+}
+#else
 #ifdef _WIN32
 #include <intrin.h>
-/* TODO: This isn't correct for many platforms */
-#ifdef _WIN64
+#ifdef _WIN64 && _M_AMD64
```

Comment:
Sorry, I should have been more explicit: `#if defined(_WIN64) && defined(_M_AMD64)` which will work even if  they have values that will fail for `&&` like `2` and `1`: `2 && 1` -> 0